tests/applications/md_grids/test_mdg_library.py

```diff
@@ -113,6 +115,67 @@ def test_two_intersecting_custom_values(self):
 
         self.check_fracture_coordinates([cc0, cc1], [fc0, fc1])
 
+    @pytest.mark.parametrize("mesh_type", ["simplex", "cartesian"])
+    @pytest.mark.parametrize("num_fractures", [1, 2])
+    def test_two_intersecting_nonmatching(self, mesh_type, num_fractures):
+        """Test meshing of a 2d square domain to generate a non-matching grid.
+
+        This is not a very powerful test, but it does verify that the generated grids
+        have  different numbers of cells in the fractures and along the mortars, and
+        that the mortar grids have a different number of grids than the matrix faces
+        tagged as on a fracture.
+
+        Parameters:
+            mesh_type: The type of mesh to generate, either 'simplex' or 'cartesian'.
+            num_fractures: The number of fractures to include in the domain.
+
+        """
+        meshing_arguments = {"cell_size": 1 / 4}
+
+        # Define the fracture endpoints and indices according to the input.
+        fracture_endpoints = []
+        fracture_indices = []
+        if num_fractures > 0:
+            fracture_endpoints.append(np.array([1 / 4, 3 / 4]))
+            fracture_indices.append(0)
+        if num_fractures > 1:
+            fracture_endpoints.append(np.array([0, 1]))
+            fracture_indices.append(1)
+
+        # Generate
+        self.mdg, _ = pp.mdg_library.square_with_orthogonal_fractures(
+            mesh_type,
+            meshing_arguments,
+            fracture_indices=fracture_indices,
+            fracture_endpoints=fracture_endpoints,
+            non_matching=True,
+            **{"interface_refinement_ratio": 2, "fracture_refinement_ratio": 4},
+        )
+
+        # Number of faces in the 2d grid that are tagged as fracture faces.
+        num_fracture_faces_from_matrix = (
+            self.mdg.subdomains(dim=2)[0].tags["fracture_faces"].sum()
+        )
+
+        # Loop over the fractures, fetch the projections to the 2d grid, and count the
+        # number of non-zero entries in the projection matrix.
+        non_zero_projection_primary = 0
+        for mg in self.mdg.interfaces(dim=1):
+            proj_primary = mg.mortar_to_primary_avg()
+            non_zero_projection_primary += proj_primary.nnz
+        # A matching grid would have equality here. We have refined the mortar grid,
+        # hence there should be more items in the projection matrix than there are
+        # fracture faces in the matrix grid.
+        assert non_zero_projection_primary > num_fracture_faces_from_matrix
```

Comment:
For cartesian grids, we could check that these match the refinement ratios, right?

Reply:
Good point, this applies also the mortar-fracture ratio below. Implemented.